### PR TITLE
WsdlProjectRunner/WsdlTestSuiteRunner. Thread-safety fix.

### PR DESCRIPTION
**Pull request by kimholan**
The WsdlProjectRunner/WsdlTestSuiteRunner uses the HashSet.size/add/remove methods to detect whether the project/suite has finished running.

HashSets are unsafe for concurrent modification, which occurs in the TestRunListenerAdapter (e.g. the unsafe hashSet.remove calls).

This can lead to thread safety issues on cpus with higher core counts due to the increased concurrency.

The fix consistently fixes to runners never finishing and being stuck on some monitor on the activeRunners.hashSet. The stacktrace when the runner hangs shows the IdleConnectionMonitorThread and the Runner waiting on the monitor. The Runner is the actual cause of the testrunner not quitting due to it waiting for the notify.

### CODE REVIEW

**Comment by AlexanderNikitin-Smartbear** [comment]
Created a JIRA ticket - https://smartbear.atlassian.net/browse/SOAPUIOS-375

**Comment by AlexanderNikitin-Smartbear** [comment]
@kimholan We need your CLA to merge this - see https://www.soapui.org/developers-corner/contribute-to-soapui.html - thanks!

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java [hunk 1]
```diff
@@ -153,85 +187,125 @@ private WsdlTestSuiteRunner runTestSuite(WsdlTestSuite testSuite, boolean async)
...
+                    testRunner.getTestSuite().removeTestSuiteRunListener(internalTestRunListener);
+
+
+                } finally {
+                    lock.unlock();
+                }
+            } finally {
+                runCount.decrementAndGet();
+                if (runCount.get() == 0) {
+                    updateStatus();
```
Comment (AlexanderNikitin-Smartbear) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java:306]: The query above looks to call `updateStatus()` once at the end. But this code does not guaranties that `updateStatus()` is called once. I suggest
`if (runCount.decrementAndGet() == 0) {
        updateStatus();
}`

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java [hunk 2]
```diff
@@ -216,32 +276,59 @@ public TestSuite getTestSuite() {
...
 
-                synchronized (activeRunners) {
-                    activeRunners.notify();
+                } finally {
+                    lock.unlock();
+                }
+            } finally {
+                runCount.decrementAndGet();
+                if (runCount.get() == 0) {
+                    updateStatus();
```
Comment (AlexanderNikitin-Smartbear) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java:331]: The same case: `updateStatus()` can be called more than once.

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java [hunk 3]
```diff
@@ -54,33 +60,49 @@ public WsdlProjectRunContext createContext(StringToObjectMap properties) {
     }
 
     public void onCancel(String reason) {
-        for (TestSuiteRunner runner : activeRunners.toArray(new TestSuiteRunner[activeRunners.size()])) {
-            runner.cancel(reason);
+        try {
+            lock.lock();
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java:64]: according to https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantLock.html

> It is recommended practice to always immediately follow a call to lock with a try block, most typically in a before/after construction such as:

```
class X {
   private final ReentrantLock lock = new ReentrantLock();
   // ...

   public void m() {
     lock.lock();  // block until condition holds
     try {
       // ... method body
     } finally {
       lock.unlock()
     }
   }
 }
```

could you please rearrange the code.

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java [hunk 4]
```diff
@@ -35,15 +35,21 @@
...
+
+    private final ReentrantLock lock = new ReentrantLock();
+    private final AtomicInteger runCount = new AtomicInteger(-1);
+    private volatile boolean running;
+
     private ProjectRunListener[] listeners;
-    private Set<TestSuiteRunner> finishedRunners = new HashSet<TestSuiteRunner>();
-    private Set<TestSuiteRunner> activeRunners = new HashSet<TestSuiteRunner>();
+    private final HashSet<TestSuiteRunner> finishedRunners = new HashSet<>();
+    private final HashSet<TestSuiteRunner> activeRunners = new HashSet<>();
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java:49]: I guess for defining the fields just Set will be enough like:
`private final Set<TestSuiteRunner> finishedRunners = new HashSet<>();`
like you do it below.

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java [hunk 5]
```diff
@@ -128,6 +161,7 @@ private void runSequential(WsdlProject project, WsdlProjectRunContext runContext
     }
 
     private void updateStatus() {
+        activeRunners.clear();
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java:164]: I suggest not to mix managing runners with updating status. My proposal is 
```
if (runCount.get() == 0) {
  activeRunners.clear();
  updateStatus();
}
```
this code resides below inside InternalTestSuiteRunListener

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java [hunk 6]
```diff
@@ -135,6 +168,7 @@ private void runSequential(WsdlTestSuite testSuite, WsdlTestSuiteRunContext runC
     }
 
     private void updateStatus() {
+        activeRunners.clear();
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java:171]: The comment above is also actual here.

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java [hunk 7]
```diff
@@ -216,32 +276,59 @@ public TestSuite getTestSuite() {
...
+            lock.lock();
+            return Arrays.asList(finishedRunners.toArray(new TestCaseRunner[finishedRunners.size()]));
+        } finally {
+            lock.unlock();
+        }
     }
 
     public int getCurrentTestCaseIndex() {
-        return currentTestCaseIndex;
+        try {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java:288]: does this really make sense for this method?

Comment (IlyaAvdeev) [reply soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlTestSuiteRunner.java:288]: the same comment is actual for all other places "simple" methods.

**Comment by IlyaAvdeev** [comment]
@kimholan we won't be able to merge any changes without getting signed contributor agreement. Thank you!

**Comment by IlyaAvdeev** [comment]
we have not get signed contributor agreement and no reply on the comments added.

**Comment by kimholan** [comment]
Hey. sorry I missed this. Going to do it now. Can this merge request be reopened / or should I open a new one? (+ fill in the CLA thing).

I would like see merged once I fix the issues you mentioned, because SoapUI is locking up at random on my Threadripper due to this.

**Pull request closed without merging**